lib/JSON/Validator.pm

```diff
@@ -407,24 +408,27 @@ sub _report_schema {
 # resolve all the $ref's that we find inside JSON Schema specification.
 sub _resolve {
```

Comment:
I'm not a huge fan of this change, since `_resolve()` is called from multiple places in the code, which might result in unexpected results, dependent on when `version()` is set the first time.

Reply:
It gets tricky. The changes I made are the easiest way I found to handle all the different usecases.  This is exactly why I moved the `$self->_id_key` call at the top of `_resolve` to lower down, to inside the `ref $schema eq 'HASH'` section where we know it's safe (for hashref schemas, we have already fetched the `'$schema'` key and therefore it is safe to call `$self->$self->version` and `$self->id_key`. For the URL/filename case we need to delay calculating the version until after we've fetched the file and can look inside its contents.

Reply:
It might be very pretty, but I think we should do this:

```
diff --git a/lib/JSON/Validator.pm b/lib/JSON/Validator.pm
index e9a7927..cacf307 100644
--- a/lib/JSON/Validator.pm
+++ b/lib/JSON/Validator.pm
@@ -161,7 +161,7 @@ sub load_and_validate_schema {
   my ($self, $spec, $args) = @_;
   my $schema = $args->{schema} || SPECIFICATION_URL;
   $self->version($1) if !$self->{version} and $schema =~ /draft-0+(\w+)/;
-  $spec = $self->_resolve($spec);
+  $spec = $self->_resolve($spec, 1);
   my @errors = $self->new(%$self)->schema($schema)->validate($spec);
   confess join "\n", "Invalid JSON specification $spec:", map {"- $_"} @errors
     if @errors;
@@ -178,7 +178,7 @@ sub new {
 sub schema {
   my $self = shift;
   return $self->{schema} unless @_;
-  $self->{schema} = Mojo::JSON::Pointer->new($self->_resolve(shift));
+  $self->{schema} = Mojo::JSON::Pointer->new($self->_resolve(shift, 1));
   return $self;
 }
 
@@ -442,7 +442,7 @@ sub _report_schema {
 # _resolve() method is used to convert all "id" into absolute URLs and
 # resolve all the $ref's that we find inside JSON Schema specification.
 sub _resolve {
-  my ($self, $schema) = @_;
+  my ($self, $schema, $detect_version) = @_;
   my $id_key = $self->_id_key;
   my ($id, $resolved, @refs);
```

By doing that, we can use the `$detect_version` flag to tell _resolve() that it is in fact allowed to set `version()`.

What do you think?

Note that the diff above is not based on your changes. I apologise for that.

Reply:
In your diff, you've still got schema being extracted wrong -- it's important to not default it to SPECIFICATION_URL until after we've looked inside `$spec` to see if `$spec->{'$schema'}` is set.

So, I can see that being explicit could be better, but I'm not sure if it will help anything here -- we can only set version if the schema URL has been determined, and to do that we need to look inside the data. That's why I did that right at the beginning, at line 158, before even calling _resolve.

Reply:
The part I don't like about my patch is that it does this code twice (and in two different functions even):

    $self->version($1) if !$self->{version} and $schema->{'$schema'} and $schema->{'$schema'} =~ /draft-0+(\w+)/;

..so it would be good to do that just once... the problem is there are so many different ways the schema could be defined, there are a lot of cases to cover.  I suspect some other code needs to move around as well to make everything more consistent.